File: scraper/scrapers/ahumada.py
import re
from html import unescape
import httpx
from .base import BaseScraper, ScrapedProduct

SEARCH_URL = "https://www.farmaciasahumada.cl/on/demandware.store/Sites-ahumada-cl-Site/es_CL/Search-Show"
BASE_URL   = "https://www.farmaciasahumada.cl"

_LINK_RE  = re.compile(r'class="pdp-link"[\s\S]{0,300}?href="([^"]+)"[^>]*>\s*([^<]+)<')
_BADGE_RE = re.compile(r'class="promotion-badge-container[^"]*"[^>]*>([\s\S]{0,300}?)(?=<img\s+class="promotion-badge|</div>)')
_PRICE_RE = re.compile(r'\$([\d.]+)')


def _parse_clp(raw: str) -> float | None:
    try:
        v = float(raw.replace(".", "").replace(",", "."))
        return v if v > 100 else None
    except ValueError:
        return None


class AhumadaScraper(BaseScraper):
    slug = "ahumada"
    name = "Ahumada"

    async def search(self, query: str) -> list[ScrapedProduct]:
        results = []
        try:
            async with httpx.AsyncClient(timeout=20) as client:
                r = await client.get(
                    SEARCH_URL,
                    params={"q": query, "start": 0, "sz": 10},
                    headers={
                        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
                        "Referer": f"{BASE_URL}/",
                        "Accept": "text/html",
                    },
                    follow_redirects=True,
                )
                html = r.text

            # Dividir en tiles por el div raíz de cada producto
            tiles = re.split(r'(?=<div[^>]+product product-tile-wrapper)', html)
            seen_pids: set[str] = set()

            for block in tiles[1:]:
                pid_m = re.search(r'data-pid=["\'](\d+)["\']', block)
                if not pid_m:
                    continue
                pid = pid_m.group(1)
                if pid in seen_pids:
                    continue
                seen_pids.add(pid)

                link_m = _LINK_RE.search(block)
                if not link_m:
                    continue
                href = link_m.group(1)
                name = unescape(link_m.group(2).strip())
                if not name:
                    continue

                is_cmr = bool(re.search(r'badge_30x40_cmr_falabella', block))

                badge_m = _BADGE_RE.search(block)
                badge_price = None
                if badge_m:
                    price_m = _PRICE_RE.search(badge_m.group(1))
                    if price_m:
                        badge_price = _parse_clp(price_m.group(1))

                if is_cmr and badge_price:
                    # Buscar todos los content= del tile (>1000 = precio CLP)
                    content_vals = [int(m.group(1)) for m in re.finditer(r'content="(\d+)"', block)
                                    if int(m.group(1)) > 1000]
                    # El precio presencial es el menor content= que sea mayor al badge (CMR)
                    sale_candidates = [v for v in content_vals if v > badge_price]
                    sale_price = min(sale_candidates) if sale_candidates else None
                    price = sale_price or badge_price
                    cmr_price = badge_price if sale_price and badge_price < sale_price else None
                else:
                    price = badge_price
                    cmr_price = None

                if not price:
                    continue

                is_bio = bool(re.search(r'bioequivalent-badge', block))
                url = href if href.startswith("http") else f"{BASE_URL}{href}"

                results.append(ScrapedProduct(
                    name=name,
                    price=price,
                    cmr_price=cmr_price,
                    has_stock=True,
                    has_online_delivery=True,
                    online_url=url,
                    is_bioequivalent=is_bio,
                ))
        except Exception as e:
            print(f"[Ahumada] Error: {e}")
        return results

File: api/src/clients/ahumada.ts
import type { ScrapedProduct } from "../lib/types.js";
import { fetchWithTimeout } from "../lib/timeout.js";
import { positiveBioSignal } from "../lib/bioequivalence.js";

const BASE = "https://www.farmaciasahumada.cl";
const SEARCH = `${BASE}/on/demandware.store/Sites-ahumada-cl-Site/es_CL/Search-Show`;

function clp(str: string): number | null {
  const match = str.replace(/\./g, "").replace(",", ".").match(/\d+/);
  if (!match) return null;
  const value = parseInt(match[0], 10);
  return value > 100 ? value : null;
}

function decodeHtml(str: string): string {
  return str
    .replace(/&aacute;/g, "á").replace(/&eacute;/g, "é")
    .replace(/&iacute;/g, "í").replace(/&oacute;/g, "ó")
    .replace(/&uacute;/g, "ú").replace(/&ntilde;/g, "ñ")
    .replace(/&Aacute;/g, "Á").replace(/&Eacute;/g, "É")
    .replace(/&Iacute;/g, "Í").replace(/&Oacute;/g, "Ó")
    .replace(/&Uacute;/g, "Ú").replace(/&Ntilde;/g, "Ñ")
    .replace(/&amp;/g, "&").replace(/&nbsp;/g, " ")
    .replace(/&#(\d+);/g, (_, n: string) => String.fromCharCode(Number(n)));
}

/**
 * S-2 (SEARCH-MATCHING-QA-01, Gate 2) — detección del badge de bioequivalencia.
 *
 * Ahumada emite el CONTENEDOR `sellcondition-bioequivalent-badges` en TODOS
 * los tiles, esté vacío o no. El badge REAL solo existe cuando dentro de ese
 * contenedor aparece
 *   <div class="bioequivalent-badge-container">
 *     <img class="js-popover bioequivalent-badge" ...>
 *   </div>
 *
 * El chequeo anterior era `block.includes("bioequivalent-badge")`, que matchea
 * el contenedor vacío por substring (`...-bioequivalent-badgeS` lo contiene) y
 * marcaba el 100% de los resultados de Ahumada como bioequivalentes (medido en
 * producción 2026-08-27: 8 búsquedas, 0 ofertas con `isBioequivalent=false`).
 *
 * La corrección exige un TOKEN de clase exacto: lookbehind de comilla/espacio a
 * la izquierda (el contenedor lo falla, porque ahí viene precedido por `-`) y
 * ausencia de `[\w-]` a la derecha (el plural `-badges` lo falla). Se mantiene
 * el enfoque por regex del resto de este adaptador; no se introduce ninguna
 * librería de parsing HTML nueva.
 */
const BIOEQUIVALENT_BADGE_RE = /class="[^"]*(?<=["\s])bioequivalent-badge(?:-container)?(?![\w-])/;

export function hasBioequivalentBadge(block: string): boolean {
  return BIOEQUIVALENT_BADGE_RE.test(block);
}

export function parseAhumadaHtml(html: string): ScrapedProduct[] {
  const results: ScrapedProduct[] = [];
  const seen = new Set<string>();
  const tileRe = /class="product product-tile-wrapper[^"]*"[^>]*data-pid="(\d+)"([\s\S]+?)(?=class="product product-tile-wrapper|<div class="col-12 grid-footer|<\/body>)/g;
  let tile: RegExpExecArray | null;

  while ((tile = tileRe.exec(html)) !== null) {
    const [, pid, block] = tile;
    if (seen.has(pid)) continue;
    seen.add(pid);

    const linkMatch = block.match(/class="pdp-link"[\s\S]*?href="([^"]+)"[^>]*>\s*([^<]+)</);
    if (!linkMatch) continue;

    const href = linkMatch[1];
    const name = decodeHtml(linkMatch[2].trim());
    if (!name) continue;

    const isCmr = block.includes("badge_30x40_cmr_falabella");
    let badgePrice: number | null = null;
    const badgeMatch = block.match(/class="promotion-badge-container[^"]*"[^>]*>([\s\S]{0,300}?)(?=<img\s+class="promotion-badge|<\/div>)/);
    if (badgeMatch) {
      const priceMatch = badgeMatch[1].match(/\$([\d.]+)/);
      if (priceMatch) badgePrice = clp(priceMatch[1]);
    }

    let price: number | null;
    let cmrPrice: number | null = null;

    if (isCmr && badgePrice) {
      const contentVals = [...block.matchAll(/content="(\d+)"/g)]
        .map((m) => parseInt(m[1], 10))
        .filter((value) => value > 1000);
      const saleCandidates = contentVals.filter((value) => value > badgePrice);
      const salePrice = saleCandidates.length ? Math.min(...saleCandidates) : null;
      price = salePrice ?? badgePrice;
      cmrPrice = salePrice && badgePrice < salePrice ? badgePrice : null;
    } else {
      price = badgePrice;
    }

    if (!price) continue;

    const imageMatch = block.match(/<img[^>]+class="[^"]*tile-image[^"]*"[^>]*src="([^"]+)"|<img[^>]+src="([^"]+)"[^>]*class="[^"]*tile-image[^"]*"/);
    const imageUrl = imageMatch ? (imageMatch[1] ?? imageMatch[2] ?? null) : null;

    results.push({
      name,
      price,
      onlinePrice: null,
      cmrPrice,
      sbpayPrice: null,
      hasStock: true,
      hasOnlineDelivery: true,
      onlineUrl: href.startsWith("http") ? href : `${BASE}${href}`,
      imageUrl,
      // CF-DATA-001 (2026-08-31): el tile de Demandware no expone marca ni
      // laboratorio. Auditado sobre el HTML real (9 búsquedas, 168 tiles): no
      // hay `product-brand`, ni `data-brand`, ni `"brand"` en el payload de
      // analítica. La marca sí está en el nombre ("**Muxol** Jarabe Adulto
      // Ambroxol Clorhidrato 600 mg 100 ml") y la deriva el dominio.
      brand: null,
      manufacturer: null,
      // BIOEQUIVALENCE-DATA-QUALITY-01 (2026-08-30): el badge es evidencia
      // POSITIVA (`<img class="js-popover bioequivalent-badge"` con
      // `data-content="<span> Bioequivalente </span>"`). Su AUSENCIA no es una
      // afirmación de Ahumada de que el producto no sea bioequivalente: el
      // contenedor `sellcondition-bioequivalent-badges` se emite vacío en todos
      // los tiles y no hay ningún marcador de "no bioequivalente". Por eso la
      // ausencia es `null`, no `false`.
      isBioequivalent: positiveBioSignal(hasBioequivalentBadge(block)),
    });
  }

  return results;
}

export async function searchAhumada(query: string): Promise<ScrapedProduct[]> {
  const params = new URLSearchParams({ q: query, start: "0", sz: "24" });
  const res = await fetchWithTimeout(`${SEARCH}?${params}`, {
    headers: {
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      "Referer": `${BASE}/`,
      "Accept": "text/html",
    },
  });

  if (!res.ok) throw new Error(`Ahumada HTTP ${res.status}`);
  const html = await res.text();
  return parseAhumadaHtml(html);
}
